src/intervals/intervalcollection.jl

```diff
@@ -0,0 +1,245 @@
+
+    push!(tree, i)
+    ic.length += 1
+end
+
+
+function show(io::IO, ic::IntervalCollection)
+    const max_entries = 8
+    n_entries = length(ic)
+    println(io, "IntervalCollection with $(n_entries) intervals:")
+    if n_entries > 0
+        for (k, i) in enumerate(ic)
+            if k > 8
+                break
+            end
+            println(io, "  ", i)
+        end
+        if n_entries > max_entries
+            print(io, "  ⋮")
+        end
+    end
+end
+
+
+function length(ic::IntervalCollection)
+    return ic.length
+end
+
+
+typealias IntervalCollectionTreeIteratorState{T} IntervalTrees.IntervalBTreeIteratorState{Int64, Interval{T}, 64}
+
+
+immutable IntervalCollectionIteratorState{T}
+    i::Int # index into ordered_trees
+    tree_state::IntervalCollectionTreeIteratorState{T}
+
+    function IntervalCollectionIteratorState(i::Int)
+        return new(i)
+    end
+
+    function IntervalCollectionIteratorState(i::Int, tree_state)
+        return new(i, tree_state)
+    end
+end
+
+
+function start{T}(ic::IntervalCollection{T})
+    update_ordered_trees!(ic)
+    i = 1
+    while i <= length(ic.ordered_trees)
+        tree_state = start(ic.ordered_trees[i])
+        if !done(ic.ordered_trees[i], tree_state)
+            return IntervalCollectionIteratorState{T}(i, tree_state)
+        end
+        i += 1
+    end
+
+    return IntervalCollectionIteratorState{T}(i)
+end
+
+
+function next{T}(ic::IntervalCollection{T},
+                      state::IntervalCollectionIteratorState{T})
+    i = state.i
+    value, tree_state = next(ic.ordered_trees[i], state.tree_state)
+
+    if done(ic.ordered_trees[i], tree_state)
+        i += 1
+        while i <= length(ic.ordered_trees)
+            tree_state = start(ic.ordered_trees[i])
+            if !done(ic.ordered_trees[i], tree_state)
+                break
+            end
+            i += 1
+        end
+    end
+
+    return value, IntervalCollectionIteratorState{T}(i, tree_state)
+end
+
+
+function done{T}(ic::IntervalCollection{T},
+                      state::IntervalCollectionIteratorState{T})
+    return state.i > length(ic.ordered_trees)
+end
+
+
+immutable IntersectIterator{S, T}
+    a_trees::Vector{IntervalCollectionTree{S}}
+    b_trees::Vector{IntervalCollectionTree{T}}
+end
+
+
+immutable IntersectIteratorState{S, T}
+    i::Int # index into a_trees/b_trees.
+    intersect_iterator
+    intersect_iterator_state
+
+    function IntersectIteratorState(i::Int)
+        return new(i)
+    end
+
+    function IntersectIteratorState(i::Int, intersect_iterator,
+                                    intersect_iterator_state)
+        return new(i, intersect_iterator, intersect_iterator_state)
+    end
+end
+
+
+@doc """
+Iterate over pairs of intersecting intervals in two IntervalCollections.
+""" ->
+function intersect{S, T}(a::IntervalCollection{S}, b::IntervalCollection{T})
+    seqnames = collect(String, intersect(Set(keys(a.trees)), Set(keys(b.trees))))
```

Comment:
Instead of:

```julia
intersect(Set(keys(a.trees)), Set(keys(b.trees)))
```

You should try:

```julia
keys(a.trees) ∩ keys(b.trees)
```

I found it simpler to read and also faster